drivers/mtd/mtd_register.c

```diff
@@ -0,0 +1,516 @@
+  mtd_close,    /* close */
+  mtd_read,     /* read  */
+  mtd_write,    /* write */
+  mtd_seek,     /* seek  */
+  mtd_ioctl     /* ioctl */
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: mtd_open
+ *
+ * Description: Open the block device
+ *
+ ****************************************************************************/
+
+static int mtd_open(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct mtd_struct_s *mtd;
+  int ret = OK;
+
+  DEBUGASSERT(inode->i_private);
+  mtd = inode->i_private;
+
+  /* Increment the reference count */
+
+  ret = nxmutex_lock(&mtd->lock);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (mtd->refs == 255)
+    {
+      ret = -EMFILE;
+    }
+  else
+    {
+      mtd->refs++;
+    }
+
+  nxmutex_unlock(&mtd->lock);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: mtd_close
+ *
+ * Description: close the block device
+ *
+ ****************************************************************************/
+
+static int mtd_close(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct mtd_struct_s *mtd;
+  int ret = OK;
+
+  DEBUGASSERT(inode->i_private);
+  mtd = inode->i_private;
+
+  /* Get exclusive access */
+
+  ret = nxmutex_lock(&mtd->lock);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  /* Decrement the reference count */
+
+  if (mtd->refs == 0)
+    {
+      ret = -EIO;
+    }
+  else
+    {
+      mtd->refs--;
+    }
+
+  nxmutex_unlock(&mtd->lock);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: mtd_read
+ *
+ * Description:  Read the specified number of bytes.
+ *
+ ****************************************************************************/
+
+static ssize_t mtd_read(FAR struct file *filep, FAR char *buffer, size_t len)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct mtd_struct_s *dev;
+  FAR char *buf;
+  off_t startblock;
+  off_t offset;
+  size_t nblocks;
+  ssize_t ret;
+
+  DEBUGASSERT(inode->i_private);
+  dev = inode->i_private;
+
+  if (len < 1)
+    {
+      return 0;
+    }
+
+  ret = nxmutex_lock(&dev->lock);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (filep->f_pos >= dev->size)
+    {
+      /* End of file */
+
+      nxmutex_unlock(&dev->lock);
+      return 0;
+    }
+
+  ret = MTD_READ(dev->mtd, filep->f_pos, len, (FAR uint8_t *)buffer);
+  if (ret < 0)
+    {
+      if (ret == -ENOSYS)
+        {
+          /* Byte read not supported, use block read */
+
+          startblock = filep->f_pos / dev->blksize;
+          nblocks = (len / dev->blksize) + 1;
+          buf = kmm_zalloc(sizeof(char) * (nblocks * dev->blksize));
```

Comment:
could we preallocate the buffer to avoid alloc/free buffer for each unaligned access